### PR TITLE
Allow for "/" in the names of methods.

### PR DESCRIPTION
**Pull request by PeterKraus**
Standardized handling of "/" in method name between frequency, hessian, gradient, properties and energy.

## Description
- Use of `_cbs_gufunc` helper function consistently across `energy`, `hessian`, `gradient`, `frequency` and `properties`
- `_cbs_gufunc` throws an exception for unsupported "true" CBS calls to `frequency` and `properties`, "method/basis" syntax works
- all relevant tests pass, see below
- modified regex in `_parse_cbs_gufunc_string()` to correctly handle `sapt2+(3)/basis` calls or `method/6-311++G` calls; splitting now only happens on variants of `+ D:`.

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] Fix for #934 

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by PeterKraus** [comment]
Hmm, there seems to be an issue with temporary files being left behind in the `energy` and `gradient` calls if they're not routed through the `cbs_gufunc` routines. If I call `clean()` in between the failing tests [like so for `cbs-xtpl-energy`](https://github.com/psi4/psi4/files/1886155/cbs-xtpl-energy-mod.txt), the test passes fine.

**Comment by PeterKraus** [comment]
There are some issues with temporary files sticking around. In the current version, if I do (excerpt from `props4` test):
```
E, wfn = prop('scf/aug-cc-pvdz', properties=["GRID_ESP", "GRID_FIELD"], return_wfn=True)
E, wfn = prop('scf/cc-pvdz', properties=["GRID_ESP", "GRID_FIELD"], return_wfn=True)
```
the code works fine, but when the bases are very different:
```
E, wfn = prop('scf/sto-3g', properties=["GRID_ESP", "GRID_FIELD"], return_wfn=True)
E, wfn = prop('scf/cc-pvdz', properties=["GRID_ESP", "GRID_FIELD"], return_wfn=True)
```
I have to call clean() in between them.

**Comment by dgasmith** [comment]
The basis sets are different in both examples? I wonder why this is exclusively for prop, can you post the full error, it would be good to track down the extra temps.

**Comment by loriab** [comment]
IIRC, big-basis-followed-by-smaller-basis is ok but not the reverse.

**Comment by PeterKraus** [comment]
OK, then I suggest that for all cases where we're switching basis sets, we should do a `clean()` - it's done in the CBS routine which is used for `energy()` and `gradient()` (as the CBS routines can handle "method/basis" syntax), we might as well do it for `properties()`.

**Comment by loriab** [comment]
Wondering if it mightn't be better to just continue routing everything with a `/` to `driver_cbs._cbs_gufunc` only now with an extra kwarg `cbs_allowed=True` that is false for `def properties()`. I worry about not breaking (or complicating) recursive driver cases (where cbs and nbody function re-enter energy, hessian, properties etc. after being broken into smaller pieces (like `mp2/cc-pv[dt]z` --> `mp2/cc-pvdz`)).

**Comment by PeterKraus** [comment]
Well, that's an obvious and simple solution I should've thought about...

**Comment by PeterKraus** [comment]
The failing test is weird: it used to do FD, now it does analytical frequencies, and that's why there's a difference. Should I enforce `dertype="energy"`?

**Comment by PeterKraus** [comment]
@dgasmith, @loriab : provided the `cbs-xtpl-freq` test passes (it does on my PC), this should be good to go from my PoV. Should I run a full test suite on this?

**Comment by loriab** [comment]
So https://github.com/psi4/psi4/blob/master/tests/cbs-xtpl-freq/input.dat#L22 is properly catching the `dertype=1` again and doing Hess by findif of Grad?

**Comment by PeterKraus** [comment]
Yeah, it was getting popped from the kwargs dict and not passed into the `func()` call. I moved the pop after single-basis methods are dealt with.

File: psi4/driver/driver.py [hunk 1]
```diff
@@ -1706,10 +1703,6 @@ def frequency(name, **kwargs):
     for postcallback in hooks['frequency']['post']:
         postcallback(lowername, wfn=wfn, **kwargs)
 
-    # Reset old global basis if needed
-    if not old_global_basis is None:
-        core.set_global_option("BASIS", old_global_basis)
```
Comment (loriab) [inline psi4/driver/driver.py]: Nice to get this old inconsistency of `old_global_basis` and `lowername, new_basis = lowername.split('/')` out of here and put freq on par with gradient.

File: psi4/driver/driver_cbs.py [hunk 2]
```diff
@@ -33,6 +33,7 @@
 import math
 import sys
 import numpy as np
+import copy
```
Comment (loriab) [inline psi4/driver/driver_cbs.py]: Did you end up using `copy`?

**Comment by dgasmith** [comment]
As a tangent I would like to figure out why we need clean at the SCF level. Cannot wait to shove everything into UUID-based HDF groups.

**Comment by PeterKraus** [comment]
@loriab I'm confused about the test results. The ctest returns:
```
The following tests FAILED:
	184 - fcidump (Failed)
	359 - dft-custom-dhdf (Failed)
	363 - dft-cation (Timeout)
	364 - dft-water-dimer (Timeout)
	384 - chemps2-caspt2-n2 (Failed)
```
where the `dft-` and `fcidump` tests failing are expected as your PR that fixes them is not yet pulled in. The `chemps2/caspt-n2` test passes when I run it manually and passed in the test log file (see attached)
[testresults.log](https://github.com/psi4/psi4/files/1889030/testresults.log)

File: psi4/driver/driver_cbs.py [hunk 3]
```diff
@@ -1586,6 +1583,16 @@ def _cbs_gufunc(func, total_method_name, **kwargs):
         else:
             return ptype_value
 
+    # Drop out for props and freqs
+    elif ptype == "properties":
+        raise ValidationError("Properties: Cannot extrapolate or delta correct properties yet.")
+    elif ptype == "frequency":
+        raise ValidationError("Frequency: Cannot extrapolate or delta correct frequencies yet.")
```
Comment (dgasmith) [inline psi4/driver/driver_cbs.py]: No reason we cannot do this for `freqs`, should get around to testing that someday. Can we also add an `else` here at the very end of `ptype` is not matched?

Comment (PeterKraus) [reply psi4/driver/driver_cbs.py]: should i do:
```
else:
    pass
```
or shift everything in by one indent?

**Comment by PeterKraus** [comment]
So I have tried to set-up a good test case, and I run into the `clean()` issue again, see attachments.
[cbs-parser.txt](https://github.com/psi4/psi4/files/1889912/cbs-parser.txt)
[cbs-parser.log](https://github.com/psi4/psi4/files/1889914/cbs-parser.log)

**Comment by loriab** [comment]
On the `clean` issue, I don't see a change in behavior from your PR. The below (on master) also requires the intervening `clean()` to avoid a PSIO error. Maybe the rules of how much the user should need to call clean need rethinking, but at least they are steady (one used to have to call it after practically everything).

```
molecule ne2 {
    0 1
    Ne 0 0 0
    --
    0 1
    Ne 2 0 0
}

set  basis cc-pvdz
sapt_global = energy("sapt2+(3)", molecule=ne2)

clean()  # PSIO error without this line

# these two calls are equivalent
#cbs_string = energy("mp2/6-311++G** + D:ccsd/6-311+G")
cbs_string = energy(cbs, corl_wfn='mp2', corl_basis='6-311++G**', delta_wfn='ccsd', delta_basis='6-311+G')
compare_values(-257.50097969, cbs_string, 5, 'cbs')
```

On the caspt2, runs fine for me, too. If you want to see the error try [looking in the testing log file](http://psicode.org/psi4manual/master/build_planning.html#faq-testsoutput). But it's certainly outside the scope of this PR.

**Comment by PeterKraus** [comment]
OK, I've re-ordered the two tests so that they don't crash without `clean()`. I had hoped that this PR might have sorted out this `clean()` issue, but for true CBS method, it's unfortunately still there when switching basis sets.

The test is meant to test the "new" functionality: allowing "method/basis" syntax with `sapt2+(3)` and + D: extrapolation with augmented Pople basis sets (both previously crashed due to "+").

@dgasmith do you want to include some `property()` or `frequencies()` calls in the test too?

**Comment by dgasmith** [comment]
@PeterKraus Yea, to round things out we should probably add those single calls tests as well to ensure that we do not regress at some point in the future (although dont make this a quickest). As a note its fine to call clean in the tests.

**Comment by PeterKraus** [comment]
The `frequency()` call is already checked in `cbs-xtpl-freq` with `dertype=1`; the handling was just moved from driver to cbs_driver in this PR.

For the `properties()`, I think it'd be easier to extend one of the `props[1-4]` tests; I've extended the `props4` call to compare the already-computed V-values from SCF with global cc-pVDZ basis with ones from an explicit SCF/cc-pVDZ calculation.

**Comment by PeterKraus** [comment]
@loriab Done, I added it to the other cbs-tests at the beginning of the list.

**Pull request merged**